### ISSUE TITLE
[CN] Splitting and merging arrays does not work

### ISSUE THREAD

**Issue by peterohanley** [rems-project/cerberus#567]
All of these functions just use the first half, second half, or all of the array. None of them work. All of them clearly should. Extracts that I've tried don't help.

```
void use16(char *a);
/*@ spec use16(pointer a);
  requires take ai = each(u64 i; i >= 0u64 && i < 16u64) {Owned(array_shift(a, i))};
  ensures take ao = each(u64 i; i >= 0u64 && i < 16u64) {Owned(array_shift(a, i))};
  @*/

void use32(char *a);
/*@ spec use32(pointer a);
  requires take ai = each(u64 i; i >= 0u64 && i < 32u64) {Owned(array_shift(a, i))};
  ensures take ao = each(u64 i; i >= 0u64 && i < 32u64) {Owned(array_shift(a, i))};
  @*/

void split_merge(void)
/*@ requires true;
    ensures true;
@*/
{
    char b[32] = {0};
    use16(b);
    /*@ extract Owned<char>, 0u64; @*/
    /*@ extract Owned<char>, 16u64; @*/
}

void split_merge_2(void)
/*@ requires true;
    ensures true;
@*/
{
    char b[32] = {0};
    use16(b+16);
}

void use_split_merge(void)
/*@ requires true;
    ensures true;
@*/
{
    char b[32] = {0};
    use32(b);
    use16(b);
}

void split_merge_use(void)
/*@ requires true;
    ensures true;
@*/
{
    char b[32] = {0};
    use16(b);
    use32(b);
}
```
```
adjacent.c:17:1: error: Cannot satisfy request for resource for de-allocating. It requires merging multiple arrays.
adjacent.c:30:5: error: Missing resource for calling function use16
adjacent.c:37:1: error: Cannot satisfy request for resource for de-allocating. It requires merging multiple arrays.
adjacent.c:50:5: error: Cannot satisfy request for resource for calling function use32. It requires merging multiple arrays.

```

**Comment by podhrmic** [comment]
@septract FYI this issue is blocking verification of the secure boot (specifically [this function](https://github.com/GaloisInc/VERSE-OpenSUT/blob/a25f42eb94325f8b239c7f325a7b230c7e4e60c2/components/platform_crypto/shave_trusted_boot/firmware.c#L113)). 

@peterohanley will attempt to add lemmas to test drive the *CN escape hatch* as I was told that reasoning about subsets of arrays is outside the decidable logic (and thus CN)

**Comment by septract** [comment]
@cp526 @dc-mak thoughts on mitigating this issue? 

What if we allowed some sort of explicit sub setting of arrays as an operation. This seems a bit analogous to constructing a Rust slice type, so for some examples it seems like it should be possible without a v fancy solver. I'd be curious what the fundamental difficulty is here?

**Comment by peterohanley** [comment]
The difficulty with the memcpy spec is that it guarantees the copy actually took place. The workaround for breaking and joining array resources with lemmas works for just assuring that *something* was written.

**Comment by dc-mak** [comment]
Merging arrays requires lemmas. The simplest way to merge values ends up using quantified constraints, which we were not keen on automatically inferring and adding.
```c
/*@
lemma merge_16_16(pointer a)
requires
  take a1 = each(u64 i; i >= 0u64 && i < 16u64) {Owned(array_shift<char>(a, i))};
  take a2 = each(u64 i; i >= 16u64 && i < 32u64) {Owned(array_shift<char>(a, i))};
ensures
  take a3 = each(u64 i; i >= 0u64 && i < 32u64) {Owned(array_shift<char>(a, i))};
  // some constraints relating a1 and a2 to a3
  // or more general properties about a3 rather than values
@*/

void split_merge(void)
/*@ requires true;
    ensures true;
@*/
{
    char b[32] = {0};
    char *b2 = b;
    use16(b);
    /*@ apply merge_16_16(b2); @*/
    /*@ extract Owned<char>, 0u64; @*/
    /*@ extract Owned<char>, 16u64; @*/
}
```
You can constrain the value of the merged array as you wish, but this way one is explicitly opting into quantified constraints (which are in general undecidable) and instatiating them.

**Comment by podhrmic** [comment]
@peterohanley is the approach that @dc-mak suggested suitable for the trusted boot example?

**Comment by septract** [comment]
cc @lwli11

**Comment by peterohanley** [comment]
> @peterohanley is the approach that @dc-mak suggested suitable for the trusted boot example?

I am using my more baroque helpers that do not require passing the original base pointer, but it works. I haven't verified the lemmas but they aren't anything unusual

**Comment by septract** [comment]
@dc-mak is this completed - done, because something has changed? Or completed - won't fix?

**Comment by dc-mak** [comment]
It's not clear to me what was requested. Splitting arrays does work in the given example: the base pointer needs to be the same. Merging arrays cannot work without lemmas because of the technical limitation of SMT theory of arrays. The lemmas would need to be proved in Rocq eventually, but that should be easy to automated in that setting.

If the issue is, "splitting should be automatically done when the base pointers aren't the same" then happy to rename it and re-open.

**Comment by cp526** [comment]
To clarify: CN could, but does not currently, support automatic splitting and merging of array ownership, without lemmas. However, reasoning about the combined array would then require manual quantifier instantiation or lemmas, but extending CN this way may still be a win, especially when the verification is only about safety, not functional correctness.

**Comment by septract** [comment]
I'm going to reopen this. The issue, indeed, is that splitting and joining arrays is very painful even in trivial cases. Some automation would be good I think. 

See MM threads here for an example of the reasoning it'd be good to support: 
* Lemma examples: https://mattermost.galois.com/verse-ta1-ta2/pl/rh3det7w7bdi3jftp3dg89sfkw
* Some target code: https://mattermost.galois.com/verse-ta1-ta2/pl/73zx9y7qrjgazbcoxamqr1kjda

**Comment by septract** [comment]
See also the helper lemmas to accomplish this that @peterohanley wrote here: https://github.com/GaloisInc/VERSE-OpenSUT/blob/6a3092f9279fc119f1a047da50307a393c37f6ed/components/platform_crypto/shave_trusted_boot/cn_array_utils.h#L25

One way to mitigate this issue would be to fold some of these lemmas into a standard library / prelude and make them polymorphic over CN types. It's unfortunate that we would have to redefine these for every array type.

**Comment by dc-mak** [comment]
I'm surprised the `Split_At` lemmas are needed, though they may be for cases where the base pointer is not the same.
The below example shows extraction of segment from the middle of an array. Indeed, the extracted guard could be arbitrary _so long as the base pointer is kept the same_.

```c
#include <string.h>
#include <inttypes.h>

void mem_access(void *buf, size_t start, size_t n);
/*$
spec mem_access(pointer buf, u64 start, u64 n);
requires
    take buf_in = each (u64 i; start <= i && i < n) { Owned(array_shift<uint8_t>(buf, i))};
ensures
    take buf_out = each (u64 i; start <= i && i < n) { Owned(array_shift<uint8_t>(buf, i))};
$*/

void test_mem_access_cutoff_start(uint8_t *target, size_t size, size_t offset, size_t len)
/*$
requires
  // offset == 2u64; // make the counterexamples nicer
  // size == 4u64;   // make the counterexamples nicer
  offset < offset + len;
  offset + len < size;
  take target_in = each (u64 i; i < size)
    { Owned(array_shift<uint8_t>(target, i))};
ensures
  take target_out = each (u64 i; i < size)
    { Owned(array_shift<uint8_t>(target, i))};
$*/
{
  mem_access(target, offset, len);
  // TODO: need a mem_merge here too ...
}
```

**Comment by dc-mak** [comment]
There are separate issues here:
1. Splitting arrays should work automatically if the interval is correct regardless of the base pointer (so long as the base pointer has the correct provenance) - **this should be feasible, and will get rid of the view shift and splitting lemmas**
2a. Merging should work automatically - merging `Block` can and should happen automatically, the cost of merging `Owned` and other predicates **is adding quantified constraints to the context - is this acceptable?** 
2b. I also note that the merge lemmas stated do not add any constraints on the merged SMT arrays, so if this is acceptable, we can also do this, and put auto-merging array values behind a flag

Please state how you would like to proceed and I can make more specific issues.

**Comment by cp526** [comment]
2a: Currently CN will only let a request for an iterated resource succeed if this is the composition of an iterated resource and a bunch of individual cells. We should just let it succeed when multiple iterated resources have to be merged and then describe the value using a `forall` in the context, yes. That's a universally quantified assertion that will not be considered automatically by the solver until the user manually instantiates it or uses it in a lemma; awkwardly, it's a constraint the user *did not write*, so we will have to issue a warning making the user aware of this.

2b: don't think we need a special flag do allow this, just the user has to be made aware of the `forall`.

1: What's the use case for splitting/merging with different base pointers. Not sure yet that's easy to add that.

**Comment by dc-mak** [comment]
> 1: What's the use case for splitting/merging with different base pointers. Not sure yet that's easy to add that.

From previous comment, see view shift lemmas:

> See also the helper lemmas to accomplish this that @peterohanley wrote here: https://github.com/GaloisInc/VERSE-OpenSUT/blob/6a3092f9279fc119f1a047da50307a393c37f6ed/components/platform_crypto/shave_trusted_boot/cn_array_utils.h#L25

@septract @peterohanley may also have opinions on 2a vs 2b

**Comment by septract** [comment]
Picking this back up since I'm working on a proof that needs it again. There is a disparity here between how hard this reasoning appears like it should be be to the user, and *how difficult this actually is to get right in CN* (going by my and @peterohanley's experience, very hard!). That seems like a good candidate for more automation.

To illustrate the use-case, here's an example bit of code using the lemmas that @peterohanley wrote (taken from [here](https://github.com/GaloisInc/VERSE-OpenSUT/blob/b7c4a192f1f952ab02e044738a0f989f1802f868/components/mission_key_management/client.c#L347) in the MKM proofs): 
```
/*$ apply SplitAt_Owned_u8(buf, (u64) buf_size, (u64) pos, (u64) (buf_size - pos) ); $*/
/*$ apply ViewShift_Owned_u8(buf, buf + pos, (u64) pos, (u64) (buf_size - pos) ); $*/

int ret = write(c->fd, buf + c->pos, buf_size - (uint64_t) c->pos);

/*$ apply UnViewShift_Owned_u8(buf, buf + pos, (u64) pos, (u64) (buf_size - pos) ); $*/
/*$ apply UnSplitAt_Owned_u8(buf, (u64) buf_size, (u64) pos, (u64) (buf_size - pos) ); $*/
```

Morally, this is quite a simple operation: the `write` operation is taking a slice from the array and acting on it, returning the same slice, which is then merged into the 'hole' remaining in the array. But this was an extremely fiddly bit of reasoning to get right! And it seems like many calls to `read` / `write`, `memcpy` etc will have this shape. 

In general, the pattern is as follows: 
* We have an object in memory
* A function operates on some sub-section of the object, so the specification is defined in terms of just this sub-section 
* We use pointer arithmetic to identify the subsection to be modified 
* The function modifies just the target area
* The calling context must make sure that the modification preserves any invariants over the overall object. 

I think this pattern is very hard to support universally. But I would guess that most instances of the pattern are just the degenerate case we see in the MKM code. We identify a slice of an array, do some manipulation, and then return the new slice, along with an invariant over the values stored in that slice.

**Comment by septract** [comment]
Reflecting a bit more...

There are really two problems here: 
1. The reasoning is 'obvious', so we'd like to be able to do it automatically. But maybe the reasoning isn't obvious enough that we can do this. 
2. We don't have any way of writing a specification for `write` that encapsulates the reasoning I did above. Ideally, we'd be able to wrap up the lemmas that @peterohanley wrote in some higher-level specification. ie. we might pass a ghost parameter to write giving the base pointer, and then the specification would do the fiddly stuff of working out the intervals. 

I don't think CN is powerful enough to do this right now without changing the source code. But if we had ghost operations, we might be able to come up with a better abstraction.

**Comment by dc-mak** [comment]
I think the common case is automatable fairly well, but to my mind the issue is how to do it in a way that complements rather than torpedoes other inflight decision

* We keep the extract/focus scheme, but start using the indices as candidates for view shifts.
* Remember we want to make extract take a base pointers #311 to improve error messages and performance.
* For similar reasons, we probably don't want to mark all iterated resources as view-shift-able
* But we also want to remove the pointer-first restriction #303
* In terms of merging arrays/inferring quantified constraints, I think users should be able to opt into this too per resource/function/file

We also want to consider how we can infer extracts in common cases such as with pointer arithmetic.[^1]

[^1]: This might be doable in the general case too, if we just ask the user to tell us how to infer indices for arbitrary iteration, or notice that one of the parameters is an `array_shift<ct>(base, quantifier)`.

**Comment by septract** [comment]
Got it - it sounds like there's some redesign needed and I certainly don't understand CN deeply enough to say what exactly should be done. This isn't a blocker for now, just a pain point. But it'd be great if whatever solution you all land on could make this problem go away 

As far as making quantifier inference / merging optional, I've been wondering for a while whether CN should have a more aggressive suggestion system. What I mean is, you have two modes: 

1. Try the proof with just the annotations that are in the file, and if it passes, good news! Otherwise if it gets stuck: 
2. Try a few likely candidates for annotations that would make the proof work. Eg. often CN knows that you need ownership of a particular resource, so just try adding an `extract` that would do the right thing. Tell the user you're running in 'search' mode, and kill it after a few seconds. If you succeed, produce a message saying "hey the proof would have moved forward if you added `whatever_tactic()`" 

The advantage of this is that mode (1) is highly predictable, but for 'obvious cases', you can also take advantage of mode (2). When the proof passes, you are always in mode (1), which avoids proof instability in future. 

From this PoV, instead of halting when CN would need an array merge, you try a couple of simple merge / quantification instantiation strategies, and if you find one, then generate the right annotation to make the merge work. 

(You could call the paper "[Thinking Fast and Slow](https://en.wikipedia.org/wiki/Thinking,_Fast_and_Slow)" although maybe someone has already done that joke ...)

**Comment by bcpierce00** [comment]
(2) is related to the idea we've discussed of giving the SMT solver more information when it fails, even though this extra information may make it behave unpredictably: This unpredictability is fine as long as you do the extra step of helping the user figure out what annotation they needed to add to make the "fast path" succeed.

**Comment by bcpierce00** [comment]
(We should totally write that paper!)

**Comment by septract** [comment]
Yes, this is related to the broader question of whether we should have an 'undecidable mode' for CN, where it is allowed to pass unbounded quantifiers to the solver. Because CVC5 produces proof witnesses, in some sense it's always true that a proof in 'undecidable mode' can be mapped into a proof in 'decidable mode'. But that's only true if you assume you have a full theorem prover to replay the proof. 

It would be interesting to think what CN annotations correspond to correct proof witnesses in CVC5. Maybe there's a small set of such annotations which would cover a large number of proofs performed in 'undecidable mode'?

**Comment by bcpierce00** [comment]
Indeed.  It seems like potentially quite a rich topic.  And well aligned with what @cassiatorczon has been thinking about (and thinking about thinking about)...

**Comment by cp526** [comment]
> It would be interesting to think what CN annotations correspond to correct proof witnesses in CVC5. Maybe there's a small set of such annotations which would cover a large number of proofs performed in 'undecidable mode'?

Looking at CVC5's language of proof witnesses for this is a great idea!
